pkg/trimaran/targetloadpacking/targetloadpacking.go

```diff
@@ -0,0 +1,266 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+targetloadpacking package provides K8s scheduler plugin for best-fit variant of bin packing based on CPU utilisation around a target load
+It contains plugin for Score extension point.
+*/
+package targetloadpacking
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/francoispqt/gojay"
+	"github.com/paypal/load-watcher/pkg/watcher"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+
+	pluginConfig "sigs.k8s.io/scheduler-plugins/pkg/apis/config"
+	"sigs.k8s.io/scheduler-plugins/pkg/trimaran"
+)
+
+const (
+	metricsAgentReportingIntervalSeconds         = 60  // Time interval in seconds for each metrics agent ingestion.
+	requestMultiplier                            = 1.5 // CPU usage is predicted as 1.5*requests for containers without limits i.e. Burstable QOS.
+	httpClientTimeoutSeconds                     = 55 * time.Second
+	metricsUpdateIntervalSeconds                 = 30
+	defaultRequestsMilliCores            int64   = 1000 // Default 1 core CPU usage for containers without requests/limits i.e. Best Effort QOS.
+	defaultHostCPUThresholdPercent       float64 = 40   // Default CPU Util Threshold. Recommended to keep -10 than desired limit.
+	Name                                         = "TargetLoadPacking"
+)
+
+var (
+	requestsMilliCores      = defaultRequestsMilliCores      // Default 1 core CPU usage for containers without requests/limits i.e. Best Effort QOS.
+	hostCPUThresholdPercent = defaultHostCPUThresholdPercent // Upper limit of CPU percent for bin packing. Recommended to keep -10 than desired limit.
+	watcherAddress          = "http://127.0.0.1:2020"
+	// Exported for testing
+	WatcherBaseUrl = "/watcher"
+)
+
+type TargetLoadPacking struct {
+	handle       framework.FrameworkHandle
+	client       http.Client
+	metrics      watcher.WatcherMetrics
+	eventHandler *trimaran.PodAssignEventHandler
+}
+
+func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin, error) {
+	err := updateArguments(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	podAssignEventHandler := trimaran.New()
```

Comment:
Adding another pod event handler doesn't seem necessary to me b/c:

1. `handle` provides functions to access all nodes and each node's pods.
1. Note that not every scheduled pod goes to scheduler's bind() cycle - a pod may be specified by other schedulers, or some component that manually assigns `.sped.nodeName`.

Reply:
I am following advise from https://github.com/kubernetes-sigs/scheduler-plugins/pull/61#issuecomment-709773644. Are you saying I will be able to add a listener to `handle` separately for `.spec.nodeName` changes? 

>Note that not every scheduled pod goes to scheduler's bind() cycle - a pod may be specified by other schedulers, or some component that manually assigns .sped.nodeName

Thats why I am watching for general events -> any change in `pod.Spec` will call `onUpdate` method right? (whether different scheduler or manual component)

Reply:
Yes in the ideal case this map won't be used. It would be however of critical use when say scheduling deployments with more than one pod and burst scheduling of single pods.